db/compaction_job.cc

```diff
@@ -171,12 +171,18 @@ struct CompactionJob::SubcompactionState {
     const std::vector<FileMetaData*>& grandparents = compaction->grandparents();
 
     // Scan to find earliest grandparent file that contains key.
+    int grandparent_index_moved = 0;
     while (grandparent_index < grandparents.size() &&
            icmp->Compare(internal_key,
                          grandparents[grandparent_index]->largest.Encode()) >
                0) {
       if (seen_key) {
         overlapped_bytes += grandparents[grandparent_index]->fd.GetFileSize();
+        if (grandparent_index == 0 ||
```

Comment:
I can't really figure out the condition. The initial value of `grandparent_index` seems it shouldn't matter since we've already overlapped that file and are now considering whether to overlap the next file. 

How about something like this:
- a copy of the `grandparent_index` at the start of the function, call it `prev_grandparent_index`
- a flag, `nontrivial_file_skipped`, which indicates if we've skipped past a file of meaningful size whose index comes after `prev_grandparent_index`

Reply:
Or we can do the first bullet point and then have a counter, `file_data_skipped`, which counts the cumulative size of all files that were skipped over. Then we can base the decision to cut the file on the total data that can be saved from future compaction, rather than whether there was a single file that was big enough.